### PR TITLE
Import album extras

### LINKED ISSUE #484: Lidarr not moving extra files
No matter what I try Lidarr refuses to move extra files (.nfo, .jpg, etc.)  I have it setup in the settings and have tried with and without a period before the extension to no avail.

Is anyone else having this issue?

It would be nice if there was an option to move any files with these extension (even if it is as is and not renamed).  

The only thing that I can thing is it isn't moving them, because after renaming the music files the remaining extra files aren't a "Matching exacta file".

image

[AB#452](https://dev.azure.com/Servarr/7ab38f4e-5a57-4d70-84f4-94dd9bc5d6df/_workitems/edit/452)

### PR DESCRIPTION
**Pull request by tty418**
#### Database Migration
YES - Constraint removed from ExtraFiles table to enable for album-wide extras to be inserted.

#### Motivation
Music collections often include additional extra files for each album such as metadata (.nfo's, EAC .log files, .txt files etc), album art images, as well as .cue sheets. Currently, these files get lost during import - both during manual import in-place, as well as from the download client. 

My goal was for this feature to work seamlessly, not break existing flows, and be **on by default**. The lack of album extras 'surprises' quite a few people, and often the revelation comes a bit late (e.g. after processing half my existing music collection),

#### Description
This PR adds handling for extra files which are specific to a whole album and not a single track. The extra files themselves will not be renamed and will be kept as-is during import. 

These 'album extras' will be handled **only if renaming is enabled**. Another requirement is that an **album directory is present** in the track naming template. If these two conditions are not satisfied, importing the album extras will not be a safe operation due to conflicts with extras from other albums.

#### Implementation details (relevant for the PR)
The current extra file handling is geared towards per-track extras. Adding support for album extras is impossible without having to touch a ton of working code and redefine interfaces. The new feature has a separate implementation as it follows completely different logic. It is easier to follow and won't break anything.

The list of relevant file extensions is configurable in the UI. The defaults were changed to something more suitable for a music service.

#### Implementation todo's (draft PR):
- [x] Test migration against PostgreS
- [x] Check file naming configuration for the presence of an album dir
- [x] Retest with downloads and manual imports after refactoring
- [x] Test rename (move)
- [x] Avoid comparing source/destination paths?

#### Todos
- [x] Tests
- [x] Translation Keys (./src/NzbDrone.Core/Localization/Core/en.json)
- [ ] [Wiki Updates](https://wiki.servarr.com)

#### Issues Fixed or Closed by this PR

* Fixes #484

### CODE REVIEW

File: src/NzbDrone.Core/Extras/Files/AlbumExtraFileManager.cs [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+        private OtherExtraFile ImportSingleFile(Artist artist, int albumId, string extraFilePath, string targetDir)
+        {
+            var fileName = Path.GetFileName(extraFilePath);
+            var fileExt = Path.GetExtension(extraFilePath);
+
+            var newFileName = Path.Combine(targetDir, fileName);
+            var transferMode = _configService.CopyUsingHardlinks ? TransferMode.HardLinkOrCopy : TransferMode.Copy;
+
+            // TODO: check for readonly / downloaditem etc, adjust transferMode
+            if (extraFilePath != newFileName)
```
Comment (tty418) [inline src/NzbDrone.Core/Extras/Files/AlbumExtraFileManager.cs:149]: I would like to avoid this check, but if source and destination are the same, `TransferFile()` will throw an exception.

File: src/NzbDrone.Core/Datastore/Migration/075_relax_not_null_constraints_extra_files.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
+using FluentMigrator;
+using NzbDrone.Core.Datastore.Migration.Framework;
+
+namespace NzbDrone.Core.Datastore.Migration
+{
+    [Migration(075)]
```
[the remark this answers is not in the document]
Comment (tty418) [reply src/NzbDrone.Core/Datastore/Migration/075_relax_not_null_constraints_extra_files.cs]: Done, also rebased on develop.

File: src/NzbDrone.Core/Extras/Files/AlbumExtraFileManager.cs (hunk 1 above)
Comment (tty418) [reply src/NzbDrone.Core/Extras/Files/AlbumExtraFileManager.cs]: Changed this comparison to work on two DirectoryInfo instances, not perfect but it covers more cases :)

**Comment by tty418** [comment]
@mynameisbogdan (or anyone reviewing): as a temporary measure I'm considering adding an option to configure the list of supported extensions for the album extras. The mid-term plan is, as before, to add configuration through the UI in a follow-up PR. But until then I could enable users or people willing to test the feature specify which extensions should be handled.

The easiest thing to do would be to read an environment variable, and then override the hardcoded list:
`if (Environment.GetEnvironmentVariable("LIDARR_EXPERIMENTAL_EXTRA_FILE_EXTENSIONS") != null) { ... }`
Or use the configuration provider and inject `IOptions<LidarrExperimentalFoo>` - cleaner, but more involved and probably unnecessary for a temporary setting.

This would resolve the issue mentioned here: https://github.com/Lidarr/Lidarr/issues/484#issuecomment-1915957008

File: src/NzbDrone.Core/Datastore/Migration/076_relax_not_null_constraints_extra_files.cs [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+{
+    [Migration(076)]
+    public class relax_not_null_constraints_extra_files : NzbDroneMigrationBase
+    {
+        protected override void MainDbUpgrade()
+        {
+            Rename.Column("TrackFileId").OnTable("ExtraFiles").To("TrackFileIdLegacy");
+            Alter.Table("ExtraFiles").AddColumn("TrackFileId").AsInt64().Nullable();
+            Execute.Sql("UPDATE \"ExtraFiles\" SET \"TrackFileId\" = \"TrackFileIdLegacy\"");
+            Delete.Column("TrackFileIdLegacy").FromTable("ExtraFiles");
```
Comment (Qstick) [inline src/NzbDrone.Core/Datastore/Migration/076_relax_not_null_constraints_extra_files.cs:14]: Does AlterColumn not work correctly in this case?

```
Alter.Table("ExtraFiles").AlterColumn("TrackFileId").AsInt32().Nullable();
```

File: src/NzbDrone.Core/Extras/ExtraService.cs [hunk 4]
```diff
@@ -48,9 +52,32 @@ public class ExtraService : IExtraService,
...
+            var firstTrack = importedTracks.First();
+            var artist = firstTrack.Item.Artist;
+            var albumId = firstTrack.Item.Album.Id;
+
+            var sourceFolder = _diskProvider.GetParentFolder(firstTrack.Item.Path);
+            var albumFiles = _diskProvider.GetFiles(sourceFolder, false);
+            var filtered = albumFiles.Where(x => !importedTracks.Select(t => t.Item.Path).Contains(x));
+
+            // TODO: grab extensions from config?
+            var extensions = new List<string> { ".cue", ".jpg", ".png", ".nfo", ".log", ".txt" };
```
Comment (Qstick) [inline src/NzbDrone.Core/Extras/ExtraService.cs:75]: _configService.ExtraFileExtensions

or a new setting if we think it's needed (existing for Artist and new for Album extensions)

File: src/NzbDrone.Core/Extras/ExtraService.cs [hunk 5]
```diff
@@ -48,9 +52,32 @@ public class ExtraService : IExtraService,
...
             _extraFileManagers = extraFileManagers.OrderBy(e => e.Order).ToList();
+            _albumExtraManager = albumExtraManager;
             _logger = logger;
         }
 
+        /// <summary>
+        /// Import extra files belonging to an album. Track-specific extras will be ignored.
+        /// </summary>
+        /// <param name="importedTracks">The imported tracks in the source dir.</param>
+        /// <param name="destinationDir">Destination dir for the album extras</param>
```
Comment (Qstick) [inline src/NzbDrone.Core/Extras/ExtraService.cs:63]: Blitz the method comments, we try and avoid it with the codebase (our style)

File: src/NzbDrone.Core/Extras/Files/AlbumExtraFileManager.cs [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+            _namingConfigService = namingConfigService;
+            _diskTransferService = diskTransferService;
+            _diskProvider = diskProvider;
+            _otherExtraFileService = otherExtraFileService;
+            _logger = logger;
+        }
+
+        public IEnumerable<ExtraFile> ImportAlbumExtras(Artist artist, int albumId, IEnumerable<string> extraFilePaths, string targetDirectory)
+        {
+            var cfg = _namingConfigService.GetConfig();
```
Comment (Qstick) [inline src/NzbDrone.Core/Extras/Files/AlbumExtraFileManager.cs:46]: `cfg` to `namingConfig`

**Review by mynameisbogdan — CHANGES_REQUESTED** [review]

File: src/NzbDrone.Core/Datastore/Migration/076_relax_not_null_constraints_extra_files.cs (hunk 3 above)
Comment (tty418) [reply src/NzbDrone.Core/Datastore/Migration/076_relax_not_null_constraints_extra_files.cs]: Since SQLite doesn't support type changes I wrongly assumed that this needs to be manually handled in the migration. Apparently Fluent Migrator managed the change correctly. Fixed 👍

File: src/NzbDrone.Core/Extras/ExtraService.cs (hunk 5 above)
Comment (tty418) [reply src/NzbDrone.Core/Extras/ExtraService.cs]: Done.

File: src/NzbDrone.Core/Extras/Files/AlbumExtraFileManager.cs (hunk 6 above)
Comment (tty418) [reply src/NzbDrone.Core/Extras/Files/AlbumExtraFileManager.cs]: Done.

File: src/NzbDrone.Core/Datastore/Migration/076_relax_not_null_constraints_extra_files.cs [hunk 7]
```diff
@@ -0,0 +1,17 @@
+using FluentMigrator;
+using NzbDrone.Core.Datastore.Migration.Framework;
+
+namespace NzbDrone.Core.Datastore.Migration
+{
+    [Migration(076)]
```
[the remark this answers is not in the document]
Comment (tty418) [reply src/NzbDrone.Core/Datastore/Migration/076_relax_not_null_constraints_extra_files.cs]: Updated attribute and migration file name.

File: src/NzbDrone.Core/Extras/ExtraService.cs (hunk 4 above)
Comment (tty418) [reply src/NzbDrone.Core/Extras/ExtraService.cs:70]: I have two questions regarding this suggestion:
1. The default value currently is `srt`, which is probably coming from the TV Arrs. 
-> Shall we define a list of suitable extensions, e.g. the ones I did my tests with and hardcoded above?
2. The setting is currently off by default.
-> I was hoping that this feature can be instead on / opt out.

Especially wrt the second point - the impact of downloading dozens of albums and finding out afterwards that none of the usual extra files (nfos, covers, cue's) are there is huge. I literally had to stop my noise file acquisition for weeks until I figured out what's happening and came up with this PR. I'm still slowly going through my spreadsheet and rebuilding artist by artist.

With that said, I assumed that this should be a different setting, which does not seem to make sense. Putting this under the umbrella of "Extra Files" is fine.

Comment (Qstick) [reply src/NzbDrone.Core/Extras/ExtraService.cs:70]: 1. Yes, let's change the default to something that makes sense. SRT doesn't (I'm actually surprised it's gone this long)
2. I'm not opposed to on by default if we think that makes sense for music.

File: src/NzbDrone.Core/Datastore/Migration/078_relax_not_null_constraints_extra_files.cs [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+using NzbDrone.Core.Datastore.Migration.Framework;
+
+namespace NzbDrone.Core.Datastore.Migration
+{
+    [Migration(078)]
+    public class relax_not_null_constraints_extra_files : NzbDroneMigrationBase
+    {
+        protected override void MainDbUpgrade()
+        {
+            Alter.Table("ExtraFiles").AlterColumn("TrackFileId").AsInt64().Nullable();
```
Comment (Qstick) [inline src/NzbDrone.Core/Datastore/Migration/078_relax_not_null_constraints_extra_files.cs:11]: Is the change from int32 to int64 intentional?

Comment (tty418) [reply src/NzbDrone.Core/Datastore/Migration/078_relax_not_null_constraints_extra_files.cs]: Not really, I did it on autopilot. Reverted to Int32
Underlying column types are, according to the SQLite documentation, int64 already.

File: src/NzbDrone.Core/Extras/ExtraService.cs (hunk 4 above)
Comment (tty418) [reply src/NzbDrone.Core/Extras/ExtraService.cs]: Alright, I made the changes. The defaults are as following: 
- Import extra files on by default;
- Extra file extensions: `log, cue, nfo, jpg, jpeg, png`

Regarding sensible defaults for the extensions:
- Must-haves are `.log` and `.cue`: highly valued in at least one particular music _community_ :)
- Nice to haves: `.nfo` (release info)
- Maybe: `.pdf` and `.txt` (release-specific, some include such extra files or docs)
- Whatever: cover files (so your jpg's, png's etc), this is recoverable information

I could add all of these, but I didn't want to make the list too long. Let me know what you think.

**Comment by tty418** [comment]
It turned out that extra file imports were handled only in the simplest case. Multi-CD releases and basically anything different than a single directory for all files resulted in files not being found, or being moved to the incorrect target directory on import. 

All problems should be solved now. At the price of quite a bit of path wrangling and extra route matching code. I tried to keep things simple and readable, with the occasional comment here and there to explain why something is done in a particular way.

The tests, both setup and the way I organized them, also needed lots of adjustments. I am happy with the end result, but faking and mocking a file system with recursive file search hurts. But without the tests I would've never made it.

The unit tests cover all **supported directory structure** variants:
- tracks at root, extras at any dir under root (search recursively)
- tracks in a single subdir, extras only under subdir (otherwise its unsafe)
- tracks in multiple subdirs (CD1, CD2 etc), extras at root or any subdir (recursively)

@mynameisbogdan  @Qstick 
TL DR:
- large parts of the PR were rewritten
- importing downloads and manually / in place works:
  - High confidence that it works well on Windows / while debugging with VS
  - Cross-platform tests caught a few issues early on
  - As soon as a Docker Image is ready I'll quickly test on linux

**Comment by tty418** [comment]
> FYI just did a rebase in #4526

I rebased develop and fixed the issue reported by Sonar. I also used the opportunity to extract some of the file matching code: 
https://github.com/Lidarr/Lidarr/pull/4370/files/59b984f621f72a5516c541a61bb76740ebdacee2..9ae4b1fb16e7e81aa57c18d4fa2b1dd44f1d6bbe

Let me know if the code is hard to follow somewhere, or if you find issues in the PR :)

**Comment by tty418** [comment]
> can you squash 36 commits into a single one as well?

Squashed 👍

**Comment by rodrigorodrigo** [comment]
just coming here to say I've been using this PR for a while (ghcr.io/linuxserver-labs/prarr:lidarr-extras-for-albums) and it's been working amazingly well. +1 on reviewing and adding this feature, finally, on lidarr.

File: src/NzbDrone.Test.Common/TestBase.cs [hunk 9]
```diff
@@ -103,7 +103,7 @@ protected string TempFolder
         [SetUp]
         public void TestBaseSetup()
         {
-            GetType().IsPublic.Should().BeTrue("All Test fixtures should be public to work in mono.");
+            GetType().Should().Match(t => t.IsPublic || t.IsNestedPublic, "All Test fixtures should be public to work in mono.");
```
Comment (augustuen) [inline src/NzbDrone.Test.Common/TestBase.cs:106]: Why is this included in this PR?

**Comment by georgedobrev** [comment]
Any movement on adding this PR to DEV?
